src/lib/control_allocation/control_allocation/ControlAllocationSequentialDesaturationTest1.cpp

```diff
@@ -0,0 +1,210 @@
+
+using namespace matrix;
+
+TEST(ControlAllocationSequentialDesaturationTest, AllZeroCase)
+{
+	ControlAllocationSequentialDesaturation control_allocation;
+	EXPECT_EQ(control_allocation.getActuatorSetpoint(), (Vector<float, ControlAllocation::NUM_ACTUATORS>()));
+	control_allocation.allocate();
+	EXPECT_EQ(control_allocation.getActuatorSetpoint(), (Vector<float, ControlAllocation::NUM_ACTUATORS>()));
+}
+
+TEST(ControlAllocationSequentialDesaturationTest, SetGetActuatorSetpoint)
+{
+	ControlAllocationSequentialDesaturation control_allocation;
+	float actuator_setpoint_array[ControlAllocation::NUM_ACTUATORS] = {1.f, 2.f, 3.f, 4.f, 5.f, 6.f};
+	Vector<float, ControlAllocation::NUM_ACTUATORS> actuator_setpoint(actuator_setpoint_array);
+	control_allocation.setActuatorSetpoint(actuator_setpoint);
+	EXPECT_EQ(control_allocation.getActuatorSetpoint(), actuator_setpoint);
+}
+
+// Make protected updateParams() function available to the unit test to change airmode after initialization
+class TestControlAllocationSequentialDesaturation : public ::ControlAllocationSequentialDesaturation
+{
+public:
+	void updateParams() { ControlAllocationSequentialDesaturation::updateParams(); }
+};
+
+class ControlAllocationSequentialDesaturationTestQuadX : public ::testing::Test
+{
+public:
+	void SetUp() override
+	{
+		// Quadrotor x geometry
+		ActuatorEffectivenessRotors::Geometry quadx_geometry{};
+		quadx_geometry.num_rotors = 4;
+		quadx_geometry.rotors[0].position = Vector3f(1.f, 1.f, 0.f); // clockwise motor numbering
+		quadx_geometry.rotors[1].position = Vector3f(-1.f, 1.f, 0.f);
+		quadx_geometry.rotors[2].position = Vector3f(-1.f, -1.f, 0.f);
+		quadx_geometry.rotors[3].position = Vector3f(1.f, -1.f, 0.f);
+		quadx_geometry.rotors[0].moment_ratio = 1.f;
+		quadx_geometry.rotors[1].moment_ratio = -1.f;
+		quadx_geometry.rotors[2].moment_ratio = 1.f;
+		quadx_geometry.rotors[3].moment_ratio = -1.f;
+
+		for (int i = 0; i < 4; ++i) {
+			quadx_geometry.rotors[i].axis = Vector3f(0.f, 0.f, -1.f); // thrust downwards
+			quadx_geometry.rotors[i].thrust_coef = 1.f;
+			quadx_geometry.rotors[i].tilt_index = -1;
+		}
+
+		// Compute actuator effectiveness
+		ActuatorEffectiveness::Configuration actuator_configuration{};
+		int num_actuators = ActuatorEffectivenessRotors::computeEffectivenessMatrix(quadx_geometry,
+				    actuator_configuration.effectiveness_matrices[0],
+				    actuator_configuration.num_actuators_matrix[0]);
+		actuator_configuration.actuatorsAdded(ActuatorType::MOTORS, num_actuators);
+
+		// Load effectiveness into allocation
+		_control_allocation.setEffectivenessMatrix(actuator_configuration.effectiveness_matrices[0],
+				actuator_configuration.trim[0], actuator_configuration.linearization_point[0],
+				actuator_configuration.num_actuators_matrix[0], true /*update_normalization_scale*/);
+	}
+
+	void setAirmode(const int32_t mode)
+	{
+		param_control_autosave(false); // Disable autosaving parameters to avoid busy loop in param_set()
```

Comment:
I'd move that to the `SetUp` method

Reply:
Makes sense.
Changed with interactive rebase: https://github.com/PX4/PX4-Autopilot/compare/24808d96cd9acebfcfb57a4f56c2e13a3f41fe57..29bf7b47b8ccc8a389ac80ad76704cf8126d9f9d